File: python/p014.py
# Longest Collatz sequence
#
# Which starting number, under one million, produces the longest chain?

collist = [0]*1000000

def collatz(n):
    orin = n
    count = 0

    while n >= orin:
        if n % 2 == 0:
            count += 1
            n //= 2
        else:
            count += 1
            n = n*3+1
    collist[orin] = count + collist[n]
    return collist[orin]

a = [1, 1]
collist[1] = 1
for i in range(2,1000000):
    c = collatz(i)
    if c > a[1]:
        a = [i, c]

print(a)

# output: [837799, 525]
# TODO: buat lebih efisien

File: go/p014.go
package main

import "fmt"

var collslice [1000000]int

func collatz(n int64) int {
	orin := n
	count := 0

	for n >= orin {
		switch n & 1 {
		case 0:
			count++
			n = n >> 1
		case 1:
			count++
			n = n<<1 + n + 1
		}
	}
	collslice[orin] = count + collslice[n]
	return collslice[orin]
}

func main() {
	var a int64 = 1
	var na int = 1
	collslice[1] = 1
	for i := int64(2); i < 1000000; i++ {
		ni := collatz(i)
		if na <= ni {
			a, na = i, ni
		}
	}
	fmt.Println(a, na)
	// output: 837799 525
}
